File: Day247_2025/Q247_GFG.java
class Node {
    int data;
    Node next;
    Node prev;

    Node(int data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

class Q247_GFG {
    private Node getKthNode(Node head, int k) {
        k -= 1;
        while (head != null && head.next != null && k > 0) {
            k--;
            head = head.next;
        }
        return head;
    }

    private Node reverseAList(Node head) {
        Node prev = null, curr = head, next = head;
        while (curr != null) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }

    public Node reverseKGroup(Node head, int k) {
        Node temp = head;
        Node prevLast = null;

        while (temp != null) {
            Node kthNode = getKthNode(temp, k);
            if (kthNode == null) {
                if (prevLast != null) {
                    prevLast.next = temp;
                }
                break;
            }

            Node nextNode = kthNode.next;
            kthNode.next = null;

            reverseAList(temp);

            if (temp == head) {
                head = kthNode;
            } else {
                prevLast.next = kthNode;
            }

            prevLast = temp;
            temp = nextNode;
        }
        return head;
    }
}

File: Day247_2025/Q247_GFG.py
class Solution:
    def getKthNode(self, head, k):
        k -= 1
        while head and head.next and k > 0:
            k -= 1
            head = head.next
        return head

    def reverseAList(self, head):
        prev, curr = None, head
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        return prev

    def reverseKGroup(self, head, k):
        temp = head
        prevLast = None

        while temp:
            kthNode = self.getKthNode(temp, k)
            if not kthNode:
                if prevLast:
                    prevLast.next = temp
                break

            nextNode = kthNode.next
            kthNode.next = None

            self.reverseAList(temp)

            if temp == head:
                head = kthNode
            else:
                prevLast.next = kthNode

            prevLast = temp
            temp = nextNode

        return head
